src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt

```diff
@@ -173,17 +175,77 @@ open class RtpPacket(
         get() = "type=RtpPacket len=$payloadLength " +
             "hashCode=${buffer.hashCodeOfSegment(payloadOffset, payloadOffset + payloadLength)}"
 
+    private var pendingHeaderExtensions: MutableList<HeaderExtension>? = null
+
+    private fun getEncodedHeaderExtension(extensionId: Int): HeaderExtension? {
+        if (!hasEncodedExtensions) return null
+
+        encodedHeaderExtensions.forEach { ext ->
+            if (ext.id == extensionId) {
+                return ext
+            }
+        }
+        return null
+    }
+
+    var hasExtensions: Boolean
+        get() = pendingHeaderExtensions?.isNotEmpty() ?: hasEncodedExtensions
+        set(value) {
+            val p = pendingHeaderExtensions
+            if (p != null) {
+                if (value && p.isEmpty()) {
+                    throw java.lang.IllegalStateException(
+                        "Cannot set hasExtensions to true with empty pending extensions"
+                    )
+                }
+                if (!value) {
+                    p.clear()
+                }
+            } else {
+                hasEncodedExtensions = value
+            }
+        }
+
     fun getHeaderExtension(extensionId: Int): HeaderExtension? {
-        if (!hasExtensions) return null
+        val activeHeaderExtensions = pendingHeaderExtensions?.iterator() ?: encodedHeaderExtensions
 
-        headerExtensions.forEach { ext ->
+        activeHeaderExtensions.forEach { ext ->
             if (ext.id == extensionId) {
                 return ext
             }
         }
         return null
     }
 
+    private fun createPendingHeaderExtensions(removeIf: ((HeaderExtension) -> Boolean)?) {
+        if (pendingHeaderExtensions != null) {
+            return
+        }
+        pendingHeaderExtensions = ArrayList<HeaderExtension>().also { l ->
+            encodedHeaderExtensions.forEach {
+                if (removeIf == null || !removeIf(it)) {
+                    l.add(PendingHeaderExtension(it))
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes the header extension (or all header extensions) with the given ID.
+     */
+    fun removeHeaderExtension(id: Int) {
+        pendingHeaderExtensions?.removeIf { h -> h.id == id }
+            ?: createPendingHeaderExtensions { h -> h.id == id }
+    }
+
+    /**
+     * Removes all header extensions except those with ID values in [retain]
+     */
+    fun removeHeaderExtensionsExcept(retain: Set<Int>) {
+        pendingHeaderExtensions?.removeIf { h -> !retain.contains(h.id) }
+            ?: createPendingHeaderExtensions { h -> !retain.contains(h.id) }
```

Comment:
This may re-encode the extensions needlessly if `retain` contains all encoded extensions, and this may be ea common case if we use this to clear non-signalled extensions in jmt. Should it be optimized to reset `pendingHeaderExtensions` to `null` if `pendingExtensions.length == encodedExtensions.length`?

Reply:
Well, I'm planning to use it to remove all hop-by-hop extensions in jmt (see the jmt PR), so in the normal cases we'll have removed something (i.e. tcc or abs-send-time).

`encodedExtensions` is an `Iterator<HeaderExtension>` not a `Collection` so it won't have a `length`, but a `removedSomething` flag is certainly doable.

Reply:
Hm, I tried writing this, but the issue is that if you did a `removeHeaderExtensionsExcept` which ends up being a no-op, followed by an `addHeaderExtension`, you'd end up creating the pending header extension vector twice.

Since for now I'd anticipate that we'd always remove a hop-by-hop header (a tcc or abs-send-time), I think not doing this optimization is fine.

Reply:
Okay. I was going to suggest swapping the extension stripper and TCC tagger to avoid re-encoding, but I guess we can and should optimize only if it proves necessary.